trin-core/src/portalnet/overlay_service.rs

```diff
@@ -469,30 +559,37 @@ where
     }
 
     /// Main bucket refresh lookup logic
-    fn bucket_refresh_lookup(&self) {
+    fn bucket_refresh_lookup(&mut self) {
         // Look at local routing table and select the largest 17 buckets.
         // We only need the 17 bits furthest from our own node ID, because the closest 239 bits of
         // buckets are going to be empty-ish.
-        let buckets = self.kbuckets.read();
-        let buckets = buckets.buckets_iter().enumerate().collect::<Vec<_>>();
-        let buckets = &buckets[256 - EXPECTED_NON_EMPTY_BUCKETS..];
-
-        // Randomly pick one of these buckets.
-        let target_bucket = buckets.choose(&mut rand::thread_rng());
-        match target_bucket {
-            Some(bucket) => {
-                let target_bucket_idx = u8::try_from(bucket.0);
-                if let Ok(idx) = target_bucket_idx {
-                    // Randomly generate a node ID that falls within the target bucket.
-                    let target_node_id = self.generate_random_node_id(idx);
-                    // Do the random lookup on this node ID.
-                    self.send_recursive_findnode(&target_node_id);
-                } else {
-                    error!("Unable to downcast bucket index.")
+        let target_node_id = {
```

Comment:
I tried breaking this out into a separate function and it ended up being up a bit much because I could no longer just `return` if anything went wrong, I needed to return an error, which this function then had to handle. That seemed like a lot for a piece of functionality only used in one place.

The reason I use the nested setup in the first place is to avoid conflict with the borrow checker: I have `self` borrowed immutably, and then `init_find_nodes_query` has to borrow it immutably. So I need the immutable `self` references to go out of scope before I call `init_find_nodes_query`.